pkg/git/provider_ado.go

```diff
@@ -0,0 +1,112 @@
+package git
+
+import (
+	"context"
+	"fmt"
+	"github.com/microsoft/azure-devops-go-api/azuredevops"
+	ado "github.com/microsoft/azure-devops-go-api/azuredevops/git"
+	"net/url"
+	"strings"
+	"time"
+)
+
+//go:generate mockery --name Ado* --output ado/mocks --case snake
+type (
+	AdoClient interface {
+		CreateRepository(context.Context, ado.CreateRepositoryArgs) (*ado.GitRepository, error)
+	}
+
+	AdoUrl interface {
+		GetProjectName() string
+	}
+
+	adoGit struct {
+		adoClient AdoClient
+		opts      *ProviderOptions
+		adoUrl    AdoUrl
+	}
+
+	adoGitUrl struct {
+		loginUrl     string
+		subscription string
+		projectName  string
+		repoName     string
+	}
+)
+
+const Azure = "azure"
+const AzureHostName = "dev.azure"
+const timeoutTime = 10 * time.Second
+
+func newAdo(opts *ProviderOptions) (Provider, error) {
+	adoUrl, err := parseAdoUrl(opts.Host)
+	if err != nil {
+		return nil, err
+	}
+	connection := azuredevops.NewPatConnection(adoUrl.loginUrl, opts.Auth.Password)
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutTime)
+	defer cancel()
+	// FYI: ado also has a "core" client that can be used to update project, teams, and other ADO constructs
+	gitClient, err := ado.NewClient(ctx, connection)
+	if err != nil {
+		return nil, err
+	}
+
+	return &adoGit{
+		adoClient: gitClient,
+		opts:      opts,
+		adoUrl:    adoUrl,
```

Comment:
I don't think you should use the parsed url here because it is parsed from `opts.Host` which is not the full url. Isn't it a problem?

Reply:
The ADO client doesn't actually need the URL. You log into ADO using the organization URL. Then via their structs provide the project and repo name. Since I use the repo name from the `CreateRepoOptions`, there were never any issues. I removed the `repoName` parsing since it wasn't actually doing anything. 

It is very confusing what is going on with the URL... any suggestions?

Reply:
How to you actually parse the project name from the `opts.Host`. Seems like it would not actually be there..

Reply:
Oh, it is :). 
https://github.com/argoproj-labs/argocd-autopilot/blob/7a325d8f5ff187ea25344b3a6efd68f9fd111a32/pkg/util/repospec.go#L30
```
		host = normalizeGitHostSpec(n[:index+len(gitDelimiter)])
```
Host = `https://dev.azure.com/TEST/rumstead/_git/` where TEST is the subscription and rumstead is the project.